### PR TITLE
DataSourceAPI: Add adhoc filters to DataQueryRequest and make it not depend on global templateSrv

### PR DESCRIPTION
**Pull request by torkelo**
### The Problem PR is trying to address
Goal is to stop needing to rely on a global adhoc filter set, to support local filters (for split views) and to have more control over which requests the filters should be applied to (instead of always ALL for the same data source). 

Changes
* Explores adding adhocFilters to DataQueryRequest (DataSourceAPI.query method arg) 
* Add adhocFilters arg to interpolateVariablesInQueries 
* Add adhocFilters arg to applyTemplateVariables

### CODE REVIEW

File: packages/grafana-data/src/types/datasource.ts [hunk 1]
```diff
@@ -552,6 +556,9 @@ export interface DataQueryRequest<TQuery extends DataQuery = DataQuery> {
   panelId?: number;
   dashboardUID?: string;
 
+  /** Filters to apply to all queries */
+  adhocFilters?: AdHocVariableFilter[];
```
Comment (dprokop) [inline packages/grafana-data/src/types/datasource.ts:560]: Just a thought- maybe we could just name it `filters` to decouple from the feature name (ad hoc filters)?

Comment (torkelo) [reply packages/grafana-data/src/types/datasource.ts:560]: @dprokop don't we want them to be coupled? so it's clear that AdHocFiltersSet results in this array? makes the transition a bit more intuitive (templateSrv.getAdhocFilters => request.adhocFilters). 

the code comments for getTagKeys and getTagValues mention adhoc filters, so it's the term we use everywhere. but I like filters as well (shorter and easier to type) so no strong feelings either way.

Comment (torkelo) [reply packages/grafana-data/src/types/datasource.ts]: @dprokop 
after thinking more I renamed property (and function args) to just filters

**Review by dprokop — APPROVED** [review]

**Comment by samjewell** [comment]
hey team, @torkelo @dprokop 

you've deprecated `getAdhocFilters` on `TemplateSrv` and replaced with a `filters` argument on `applyTemplateVariables` / `DataQueryRequest`. That migration path works well for the query execution path.

However, there is a plugin UI context — specifically inside a query editor — where I've been using the current AdHoc filters but there's no `DataQueryRequest` available. The only option today is the deprecated `(templateSrv as any).getAdhocFilters()`, which:

1. Isn't on the public `TemplateSrv` interface (requires `as any` cast)
2. Fires a deprecation warning at runtime
3. Fails TypeScript compilation if called without casting

### Concrete example

image


The [Cube datasource plugin](https://github.com/grafana/grafana-cube-datasource): Our query editor shows a SQL preview, and it includes the current AdHoc filters. I'm currently migrating away from `getAdhocFilters` ([PR #129](https://github.com/grafana/grafana-cube-datasource/pull/129)), and the request-path side is straightforward, but the query editor preview has no clean replacement:

- [QueryEditor.tsx (line 22-25)](https://github.com/grafana/grafana-cube-datasource/blob/03db83168e4913420c524b0a5a718c1819950ba3/src/components/QueryEditor.tsx#L22-L25) — where we need AdHoc filters in the editor context
- [buildCubeQuery.ts (line 15-18)](https://github.com/grafana/grafana-cube-datasource/blob/03db83168e4913420c524b0a5a718c1819950ba3/src/utils/buildCubeQuery.ts#L15-L18) — the function that builds the query preview and now accepts filters as an argument

### Question

Is there a recommended non-deprecated way for plugin query editors to subscribe to / read current AdHoc filter state? For example, via the scenes framework, or a hook, or something on the `QueryEditorProps`?

If not, would you be open to either:
- Exposing `getAdhocFilters` (or equivalent) on the public `TemplateSrv` interface, or
- Providing a React hook like `useAdHocFilters(datasourceName)` for plugin editor contexts?

Or do you have a strong opinion here around the code, or UX, or both, that I mustn't show the full query **including filters** that will run, when the query is executed? Do you know what competitor products do here? I'm personally most interested in those within the BI space, for example Looker.

**Pull request merged**